server/src/main/java/io/crate/execution/engine/distribution/DistributingConsumer.java

```diff
@@ -140,40 +142,47 @@ private void consumeIt(BatchIterator<Row> it) {
     private void forwardFailure(@Nullable final BatchIterator it, final Throwable f) {
         Throwable failure = SQLExceptions.unwrap(f); // make sure it's streamable
         AtomicInteger numActiveRequests = new AtomicInteger(downstreams.size());
-        DistributedResultRequest request =
-            new DistributedResultRequest(jobId, targetPhaseId, inputId, bucketIdx, failure, false);
+
         for (int i = 0; i < downstreams.size(); i++) {
             Downstream downstream = downstreams.get(i);
             if (downstream.needsMoreData == false) {
                 countdownAndMaybeCloseIt(numActiveRequests, it);
             } else {
                 if (traceEnabled) {
                     LOGGER.trace("forwardFailure targetNode={} jobId={} targetPhase={}/{} bucket={} failure={}",
-                        downstream.nodeId, jobId, targetPhaseId, inputId, bucketIdx, failure);
+                                 downstream.nodeId, jobId, targetPhaseId, inputId, bucketIdx, failure);
                 }
-                distributedResultAction.pushResult(downstream.nodeId, request, new ActionListener<>() {
-                    @Override
-                    public void onResponse(DistributedResultResponse response) {
-                        downstream.needsMoreData = false;
-                        countdownAndMaybeCloseIt(numActiveRequests, it);
-                    }
+                distributedResultAction.execute(
+                    DistributedResultRequest.of(downstream.nodeId,
```

Comment:
Could change this so that the failure request is still only created once outside of the loop and that only the wrapper is created inside. Would avoid a couple of allocations.